src/vulkan/pipeline.cc

```diff
@@ -65,75 +65,60 @@ void Pipeline::Shutdown() {
   if (r.IsSuccess())
     command_->SubmitAndReset(fence_timeout_ms_);
 
-  DestoryDescriptorPools();
-  DestoryDescriptorSetLayouts();
+  for (auto& info : descriptor_set_info_) {
+    vkDestroyDescriptorSetLayout(device_, info.layout, nullptr);
+    if (info.empty)
+      continue;
+
+    vkDestroyDescriptorPool(device_, info.pool, nullptr);
+
+    for (auto& desc : info.descriptors_)
+      desc->Shutdown();
+  }
+
   command_->Shutdown();
   vkDestroyPipelineLayout(device_, pipeline_layout_, nullptr);
-  for (size_t i = 0; i < descriptors_.size(); ++i)
-    descriptors_[i]->Shutdown();
   vkDestroyPipeline(device_, pipeline_, nullptr);
 }
 
-void Pipeline::DestoryDescriptorSetLayouts() {
-  for (size_t i = 0; i < descriptor_set_layouts_.size(); ++i) {
-    vkDestroyDescriptorSetLayout(device_, descriptor_set_layouts_[i], nullptr);
-  }
-  descriptor_set_layouts_.clear();
-}
-
 Result Pipeline::CreateDescriptorSetLayouts() {
-  std::sort(descriptors_.begin(), descriptors_.end());
+  for (auto& info : descriptor_set_info_) {
+    VkDescriptorSetLayoutCreateInfo desc_info = {};
+    desc_info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_LAYOUT_CREATE_INFO;
 
-  size_t i = 0;
-  while (i < descriptors_.size()) {
+    // If two descriptor sets are discontinuous, we must put empty
```

Comment:
I think this comment is out of date. Something like:

If there are no descriptors for this descriptor_set we only need to create the layout and
there will be no bindings.

Reply:
done